test/rule_creation_protocol/protocol.py

```diff
@@ -0,0 +1,301 @@
+        """ The rule code is made of 3 digits capital letters (taken from the list of Functional parts)
+        + 3 digit numbers 
+        e.g. 'ALB123', 'GEM391' and not 'AB14' or 'GKE9001'
+        """
+        validator_helper = ValidatorHelper()
+        validation = validator_helper.validate_rule_code(value)
+        if validation:
+            raise ProtocolError(
+                value=validation['value'],
+                message=validation['message']
+            )
+        return value
+
+    @field_validator('first_separator')
+    def validate_first_separator(cls, value, values):
+        valid_first_separator = values.data['valid_first_separator']
+        if value != valid_first_separator:
+            raise ProtocolError(
+                value=value,
+                message=f"expected separator {valid_first_separator} but instead found {value}"
+            )
+        return value, values
+
+    @field_validator('separators')
+    def validate_separators(cls, value, values):
+        separators = value
+        valid_separators = values.data['valid_separators']
+        unvalid_separators = [separator for separator in separators if separator != valid_separators]
+        if any(unvalid_separators):
+            raise ProtocolError(
+                value = separators,
+                message = f"expected {valid_separators} but found the following unvalid seperators {unvalid_separators}"
+            )
+        return value, values
+
+
+class RuleCreationConventions(ConfiguredBaseModel):
+    """Validate feature conventions
+    """
+    feature: Naming  # feature as in .feature file
+    dotfeature_file: Naming  # either user input or test file
+    ifc_input: dict
+    tags: list
+    description: str
+    steps: list
+    filename: str
+    readme: str
+
+    @field_validator('tags')
+    def do_validate_tags(cls, value) -> dict:
+        validator = ValidatorHelper()
+        validated_tags = validator.validate_tags(value)
+        if validated_tags != 'passed':
+            raise ProtocolError(
+                value=validated_tags['value'],
+                message=validated_tags['message']
+            )
+        return value
+
+    @field_validator('ifc_input')
+    def validate_ifc_input(cls, value):
+        # @todo implement
+        pass
+
+    @field_validator('description')
+    def validate_description(cls, value=list) -> list:
+        """must include a description of the rule that start with "The rule verifies that..."""  # allow for comma's
+        if not any(value.startswith(f"{prefix} rule verifies{optional_comma} that") for prefix in ("This", "The") for optional_comma in ("", ",")):
+            raise ProtocolError(
+                value=value,
+                message=f"The description must start with 'The rule verifies that', it now starts with {value}"
+            )
+        return value
+
+    @field_validator('steps')
+    def validate_steps(cls, value):
```

Comment:
There are multiple "rules" checked here within one validate_steps method. While it has the drawback, that not all (only the first) step errors are listed at once per run, I am ok with this. It seems elegant to have all the steps check in one place. 

What do you think @Ghesselink ?

Reply:
I'm not sure if I understand. `value` here is a list of step dictionaries and all the steps within a rule are tested here.
`all(d['keyword'] in ['Given', 'Then', 'And'] for d in value)`

for instance with ALS004
```
    Given A file with Schema Identifier "IFC4X3_TC1" or "IFC4X3_ADD1" or "IFC4X3" or "IFC4X3_ADD2"
    And An IfcAlignmentSegment
    And Its attribute Representation
    And Its attribute Representations
    Then The value of attribute RepresentationIdentifier must be Axis
    And The value of attribute RepresentationType must be Segment
    And The type of attribute Items must be IfcCurveSegment
```

```
[d['keyword'] for d in value]
['Given', 'And', 'And', 'And', 'Then', 'And', 'And']
```